### ISSUE TITLE
SIGSEGV in mbdump.c

### ISSUE THREAD

**Issue by MASHtm** [cyrusimap/cyrus-imapd#2808]
I'm reporting this for 2.5.13, but even master contains the relevant code...

We use murder backend to backend renames regularly, but recently the destination backend process crashed and the backtrace shows:
```
#0  0x00007fc01b30a9df in __strlen_sse42 () from /lib64/libc.so.6
#1  0x00007fc01da74e4b in quota_update_useds (quotaroot=0x560585f491e0 "user.xxxxxxx", diff=0x7ffd8ba94a00, mboxname=0x0) at imap/quota_db.c:474
#2  0x00007fc01da5b0d9 in undump_mailbox (mbname=<value optimized out>, pin=<value optimized out>, pout=0x560585f0ff10, auth_state=<value optimized out>)
    at imap/mbdump.c:1287
#3  0x0000560584bd1789 in cmd_undump () at imap/imapd.c:9855
#4  cmdloop () at imap/imapd.c:2153
#5  0x0000560584bd2e21 in service_main (argc=<value optimized out>, argv=<value optimized out>, envp=<value optimized out>) at imap/imapd.c:984
#6  0x0000560584bd420a in main (argc=<value optimized out>, argv=<value optimized out>, envp=0x7ffd8ba9b568) at master/service.c:606
```

The crash is clear looking at the code because
mbdump.c/undump_mailbox():1287 calls
```
            if (changed) {
                r = quota_update_useds(mailbox->quotaroot, quota_usage, 0);
```

Looking at quota_db.c calling `quota_update_useds(..., ..., 0)` MUST crash since it contains mboxname and is used without further check for `strlen(mboxname)` at ...
```
        if (q.scanmbox) {
            cmp = cyrusdb_compar(qdb, mboxname, strlen(mboxname),
                                 q.scanmbox, strlen(q.scanmbox));
        }
```

Since we usually set the mailbox to readonly and block user access before migration to a new backend it seems it took years until somebody won the race to trigger this condition that the quota needs to be updated on the destination.

I'm not sure how to fix this yet.

**Comment by elliefm** [comment]
What happens if we just change that condition to like, `if (mboxname && q.scanmbox) {`?

We've already pre-initialised `cmp` to `1`... is that a sensible default?

**Comment by elliefm** [comment]
Oh I see, once upon a time `quota_update_useds()`'s last argument was an `int` flag which said whether to do the scan behaviour.  And when the function signature changed that argument to be `const char *` I guess nobody's compiler complained about int/pointer, so none of the old code paths that used `0` for that argument got updated.

So, it seems fairly clear to me that all the places that pass `0` (or did, until I updated them to NULL) were doing so specifically to request the scan behaviour be skipped.

We now initialise `cmp = 1` and only do the scan behaviour if `cmp <= 0`, so it should do the correct thing with the change I suggested in my last comment (assuming "skip the scan behaviour" is still the right thing here).
